feat(conditions): Move the condition assignments interface to its own class

petervdmeer opened the pull request:
**Feature**
This PR makes a first step in addressing the feature described in issue #10213

## Summary
This PR makes a first step in splitting the ConditionSet into expressions and assignments; it splits the interface, and makes sure that various users of conditions call the correct version of the class (depending on if they want to query, or if they want to apply conditions).

This PR is mostly groundwork to allow introducing a parser that generates an Abstract Syntax Tree in a follow-up PR. (I plan to upload a draft of such a follow-up PR soon after this PR gets merged.)

Splitting up the interfaces also makes it easier to write more unit tests specifically for assignments, and tests specifically for testable expressions. (The arithmetic parsing code seems to be a bit low on test-coverage.) This is also for a follow-up PR.

## Screenshots
N/A

## Usage examples
For coders:
- Use ConditionSet for conditions you want to check.
- Use ConditionAssignments if you want to modify conditions.

For content-creators:
- The Wiki already has the correct descriptions of how things are supposed to work, so just use conditions as described on the wiki.

For end-users:
- End users should not notice anything from this PR.

## Testing Done
The unit-tests have been updated, and I verified that most calls map 1:1 from the old situation to the new situation.

## Save File
N/A. This can be tested with any savefile. (But the best tests are probably the unit-tests.)

## Wiki Update
The Wiki already has separate sections for `Testable condition sets` and `Applied condition sets`, so that doesn't seem to need an update.

## Performance Impact
I expect no performance impact.

RisingLeaf commented on source/ConditionSet.h:
```diff
@@ -45,9 +45,22 @@ class ConditionSet {
 	// Save a set of conditions.
 	void Save(DataWriter &out) const;
 
+	// Change this condition to always be false.
+	void MakeNever();
+
 	// Check if there are any entries in this set.
 	bool IsEmpty() const;
 
+	// Check if the given condition values satisfy this set of expressions. First applies
+	// all assignment expressions to create any temporary conditions, then evaluates.
+	bool Test(const ConditionsStore &conditions) const;
+
+	// Get the names of the conditions that are modified by this ConditionSet.
+	std::set<std::string> RelevantConditions() const;
+
+
+
+private:
```
please move the functions in the source file too

RisingLeaf reviewed (changes requested):
Might be a good idea to edit the comment on top of the ConditionSet class

petervdmeer commented:
> Might be a good idea to edit the comment on top of the ConditionSet class

Done, I also described a bit of the next change that comes after this PR.

RisingLeaf reviewed (approved):
I did not look at the test cause I have absolutely no idea about the tests

TomGoodIdea commented on source/ConditionAssignments.cpp:
```diff
@@ -0,0 +1,95 @@
+
+
+
+bool ConditionAssignments::Add(const vector<string> &lhs, const string &op, const vector<std::string> &rhs)
```
```suggestion
bool ConditionAssignments::Add(const vector<string> &lhs, const string &op, const vector<string> &rhs)
```

TomGoodIdea commented on source/ConditionSet.cpp:
```diff
@@ -280,6 +288,38 @@ bool ConditionSet::IsEmpty() const
+// Get the names of the conditions that are relevant for this ConditionSet.
+set<string> ConditionSet::RelevantConditions() const
+{
+	set<std::string> result;
```
```suggestion
	set<string> result;
```

TomGoodIdea reviewed (commented):
I'm not really sure if this split into two classes is needed for the features you mentioned. IMO it adds some complexity, and could be avoided, especially if most of the new functions are just simple operations on the underlying ConditionSet.

petervdmeer commented:
> I'm not really sure if this split into two classes is needed for the features you mentioned.

It is needed for an AST parser, because parsing for `Applied condition sets` and `Testable condition sets` (as the Wiki calls them) is different.

Having an AST parser has certain benefits, for example:
- syntax-errors can be detected at load-time. (The current ConditionSet code detects certain errors at runtime.)
- it could decide at load-time that `8 + 3 * 5` evaluates to `23`, and just immediately store `23` instead of the expression/calculation (saving calculations later at runtime).
- `someVar and never` can be optimized to `never` at load-time, removing unnecessary lookups for `someVar` at runtime.
- that we lose the overhead of the intermediate `const ConditionsStore &created` conditions when only testing/evaluating conditions.
- condition-monitoring can be implemented a bit more cleanly.
- we can return 'int64_t' results from ConditionSet, instead of only boolean. Allowing conditions to be used for more than just yes/no questions.

> IMO it adds some complexity

I feel that large part of this complexity comes from the current situation of having both `Applied condition sets` and `Testable condition sets` combined into a single class. This complexity should be gone when all support for assignment operators are moved out of ConditionSet into ConditionAssignments.

> and could be avoided, especially if most of the new functions are just simple operations on the underlying ConditionSet.

Agreed, I plan to avoid this by introducing an ASR parser (in a follow-up PR); the ASR parser code will move all assignment operators out of the ConditionSet class, into the ConditionExpression class, removing the need to forward assignment-related operations to the ConditionSet.

tibetiroka commented on source/ConditionSet.cpp:
```diff
@@ -272,6 +272,14 @@ void ConditionSet::Save(DataWriter &out) const
+void ConditionSet::MakeNever()
+{
+	// Add the equivalent "never" condition, `"'" != 0`.
+	Add("has", "'");
```
Is there anything preventing a plugin from declaring the `'` condition as true?

petervdmeer replied:
No, I don't think anything is preventing a plugin from declaring the '\'' condition to be true. This is not a new issue, this issue only moved from Mission.cpp towards here.
I already planned to solve this in a follow-up PR, and I added 2 TODO items as reminders so that it doesn't get forgotten.

tibetiroka replied:
I've been thinking. According to our data format, we can't parse
> " `

as a token. It can't be quoted in quotes nor backticks, and it has a whitespace so it must be quoted.

It's a bit hacky, but this is something plugins can't possibly override. Well, unless there are some weird edge cases within our data parsing code.

RisingLeaf replied:
Just have a check for one specific word and don't do hacks

petervdmeer replied:
> Just have a check for one specific word and don't do hacks

Or we just validate if condition-names are valid by using a regular expression when loading conditions (and their names).

petervdmeer replied:
Probably not in this PR though, but the TODO item is at least created.

tibetiroka commented on tests/unit/src/test_conditionSet.cpp:
```diff
@@ -76,22 +76,22 @@ SCENARIO( "Extending a ConditionSet", "[ConditionSet][Creation]" ) {
-			AND_THEN( "a log message is printed to assist the user" ) {
-				REQUIRE( warnings.Flush() == validationWarning );
-			}
+			// AND_THEN( "a log message is printed to assist the user" ) {
```
I'm not super happy about these disabled tests.

petervdmeer replied:
I agree, we should have tests for "empty" and "invalid" conditions, both for ConditionSets and for ConditionAssignments. I was able to restore the test for "invalid" conditions after dealing with some string beautification functionality. Will check later what I can do for the other tests.

petervdmeer replied:
I restored both checks, but some condition-users actually try to load empty conditions (which now gives an error-message). I'll update those callers not to load the condition when the condition is empty.

petervdmeer replied:
Also updated the substitutions in conversations and a branch-statement in the data-files that was missing conditions. I believe this is solved now.

tibetiroka commented on source/ConditionSet.cpp:
```diff
@@ -247,6 +247,8 @@ ConditionSet::ConditionSet(const DataNode &node)
 void ConditionSet::Load(const DataNode &node)
 {
 	isOr = (node.Token(0) == "or");
+	if(! node.HasChildren())
```
```suggestion
	if(!node.HasChildren())
```

tibetiroka commented:
That segfault is making me feel uncomfortable, especially that I can't reproduce it.

petervdmeer commented:
> That segfault is making me feel uncomfortable, especially that I can't reproduce it.

Agreed, it happened on Windows twice by now. But it doesn't seem to happen on Linux builds.

tibetiroka commented:
Unfortunately, I don't have access to a windows machine that can run the game. The best I have is a VM that can compile it, but doesn't support OpenGL properly so the game doesn't even launch.

warp-core commented on tests/unit/src/test_conditionAssignments.cpp:
```diff
@@ -0,0 +1,125 @@
+
+		THEN( "no assignments are added from empty nodes" ) {
+			const std::string validationWarning = "Error: Loading empty (sub)condition:\n";
+			set.Load(DataNode{});
```
Running the unit tests locally (built with clang) this call results in a vector subscript out of bounds error.
This empty data node has no tokens. When the node is passed to `ConditionSet::Load`, it expects at least one token, so it calls `node.Tokens(0)`, which either fails a debug assertion or results in a segfault, because there isn't actually anything there.
I would suggest that this test case should just be removed: DataFile should never produce a DataNode with no tokens, such an object has no meaning. It real DataNodes can actually be empty, we have a much bigger problem.

petervdmeer replied:
Thanks for finding this!

I changed it to be a node without any children, because that should also register as empty.

petervdmeer commented:
> That segfault is making me feel uncomfortable, especially that I can't reproduce it.

This is now fixed, with some help from warp-core.

tibetiroka commented on data/korath/korath missions.txt:
```diff
@@ -419,6 +419,7 @@ mission "Efreti: Distress Call 4"
 				to display
 					"flagship crew" <= 1
 			branch accept
+				1 == 1
```
I think the empty branch is no longer true. (Based on some discussion around a wiki pr, this wasn't a documented/expected feature to begin with, though I'm slightly worried about plugins subtly breaking.)

petervdmeer replied:
I wouldn't expect _this_ PR to change the behaviour of an empty branch (it should only introduce a warning).
But what would be the proper construct to use here? Should this just be accept without the branch? Or should this always have a false condition?

tibetiroka replied:
I think it previously defaulted to a true condition, so that's what we should continue to use.

petervdmeer replied:
I removed this change from this PR. Checking on empty conditions under branches seems to be quite a big topic that deserves it's own ticket and PR.

tibetiroka commented on source/Conversation.cpp:
```diff
@@ -160,7 +160,10 @@ void Conversation::Load(const DataNode &node)
 			nodes.emplace_back();
 			nodes.back().canMergeOnto = false;
-			nodes.back().conditions.Load(child);
+			// Let's support empty conditions for now as well.
+			// Not fully officially supported, but not fully unsupported either...
```
```suggestion
			// Maintenance note: Empty conditions might have to be removed in the future,
			// so their support is unofficial.
```

tibetiroka commented on source/ConditionAssignments.h:
```diff
@@ -0,0 +1,69 @@
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONDITION_ASSIGNMENTS_H_
+#define CONDITION_ASSIGNMENTS_H_
```
```suggestion
#pragma once
```

tibetiroka commented on source/ConditionAssignments.h:
```diff
@@ -0,0 +1,69 @@
+
+	ConditionSet setToEvaluate;
+};
+
+#endif
```
```suggestion
```

petervdmeer commented on source/ConditionSet.cpp:
```diff
@@ -272,6 +274,16 @@ void ConditionSet::Save(DataWriter &out) const
+	// Add the equivalent "never" condition, `"'" != 0`.
+	// TODO: change the ConditionSet to contain "false" literal, instead of a comparison that should return "false".
+	// TODO: Validate if condition-names are valid. (The "'" character by itself should not pass as a valid condition.)
+	Add("has", "'");
```
yes; https://github.com/endless-sky/endless-sky/blob/f0298e3354db939b5370b0b9ada9aac31a64beab/source/ConditionSet.cpp#L369

petervdmeer commented on source/ConditionSet.h:
```diff
@@ -27,10 +27,11 @@ class DataWriter;
+// A condition set is a collection of operations on the player's set of named "conditions"; "test" operations that just
+// check the values of those conditions, and in the future "evaluation" operations that can calculate an int64_t value
+// based on the conditions.
+// Operations that can be "applied" to change the values were historically also here, but are being moved to
```
Well, the actual move happens in a follow-up PR (ConditionAssignments is currently only fowarding the calls it has). So I was planning to delete this line and the next one in the follow-up PR.

petervdmeer commented:
I removed "temporary testable assignment conditions", as discussed in the context of #10406. And I moved the toplevel Apply function to ConditionAssignments, because Apply is only used in ConditionAssignments now.

tibetiroka commented:
I am, but that's not the solution we went for. It's a bit of a pain to integrate a java tool into our build process.

petervdmeer commented:
> I assume you're aware of ANTLR?

Yes, you mentioned ANTLR in https://github.com/endless-sky/endless-sky/issues/10213#issuecomment-2171429335 (the issue-ticket which was filed in preparation for this PR).
